File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxLen = 0;
        int left = 0;
        int right = 0;
        Set<Character> seen = new HashSet<>();

        while(right < s.length()){
            if(seen.contains(s.charAt(right))){
                while(seen.contains(s.charAt(right))){
                    seen.remove(s.charAt(left));
                    left++;
                }

            }

            seen.add(s.charAt(right));
            right++;
            maxLen = Math.max(maxLen, right - left);

        }

        return maxLen;

    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-4.js
class Solution {
    /**
     * @param {number[]} prices
     * @return {number}
     */
    maxProfit(prices) {
        let left = 0;
        let right = 1;
        let maxProfit = 0;

        while(right < prices.length){
            let profit = prices[right] - prices[left];

            if(profit > 0){
                maxProfit = Math.max(maxProfit, profit);
            }
            else{
                left = right;
            }

            right++;
        }

        return maxProfit;
    }
}
